### PR TITLE
Makes TypeScript a peer dependency

### LINKED ISSUE #159: Make TypeScript a peer dependency
We've just moved to TypeScript 4, but the library requires 3.8 so both are installed. It appears to work fine at the moment, but it would better if it were a peer dependency so the same version is used (TypeScript 5 syntax, for example, might break it).

### PR DESCRIPTION
**Pull request by sverweij**
## What's in here?

- Moves typescript from a regular to a development dependency
- Declares typescript as a (mandatory) peer dependency, taking the current ^-version as the minimum (although >=3.0.0 might be better for backwards compatibility?)
- Expands the travis ci config so it runs against typescript 3 and 4
- Updates the Dockerfile so it installs typescript (latest) after the npm ci.
- Also: adds an .npmrc to make sure the current npm installation & upgrade policies (^, package-lock, if I inferred correctly) are respected on machines that have a global override for these (if you rather have this one in a separate PR, let me know).

## Why?

Fixes #159 

## Test plan

- [x] green ci
- [x] manual run of npm pack (to check whether the change to the `Dockerfile` panned out all right)
- [x] installation of the tarball on a local TypeScript project + running it through its paces there

## Note

As noted in the issue this might be construed as a 💥 breaking change because there might be one or two environments that use ts-unused-exports but that don't have typescript installed in the same spot.

### CODE REVIEW

**Comment by pzavolinsky** [comment]
Hey @sverweij this is awesome!

I'm currently swamped right now so it might take me a while to go through it and I'm not sure about @mrseanryan availability, please be patient, I have all the intention of merging this change.

The only comment I have is please revert changes in `package-lock.json` as they don't really change anything.

Agreed on the breaking change, after this merge I'll publish a major version.

Thanks again for your contrib.

Cheers!

File: package-lock.json [hunk 1]
```diff
@@ -3081,7 +3081,8 @@
     "typescript": {
       "version": "3.8.3",
       "resolved": "https://registry.npmjs.org/typescript/-/typescript-3.8.3.tgz",
-      "integrity": "sha512-MYlEfn5VrLNsgudQTVJeNaQFUAI7DkhnOjdpAp4T+ku1TfQClewlbSuTVHiA+8skNBgaf02TL/kLOvig4y3G8w=="
+      "integrity": "sha512-MYlEfn5VrLNsgudQTVJeNaQFUAI7DkhnOjdpAp4T+ku1TfQClewlbSuTVHiA+8skNBgaf02TL/kLOvig4y3G8w==",
+      "dev": true
```
Comment (sverweij) [inline package-lock.json:3085]: the `"dev": true` is the only real change here - the hash remains what it is, the line just grew a trailing comma.

**Comment by sverweij** [comment]
hi @pzavolinsky thanks - totally get the swamped-ness (is that a word?), no worries; it's volunteer work so take your time. It's already great you put this project out here and put in the work to maintain it! 

On reverting changes to package-lock.json - hadn't even looked at them, just let npm cli do it's job on it. 
- I reverted the hash change on mini-match (which changed from from sha1 -> sha512 🤷 , but is now back at the sha1 it was before this PR). 
- The other change is that the typescript dependency got the `dev` attribute with value `true`. It seems that that is still necessary to stay in sync with typescript's new status as a devDependency in package.json (had to look thrice to actually see that _that_ was the actual change. I've highlighted it with a comment so it's easier to spot).

**Comment by sverweij** [comment]
Addendum on the minimatch hash - when you run `npm install` on either `master` or the branch pertaining to this PR with npm version 9 (the one that ships with node 14) it auto-updates the sha1 hash in package-lock.json to sha512. On node 12 it leaves the lock file alone, though 🎩->🐇 - 🤷.

File: package-lock.json (hunk 1 above)
Comment (mrseanryan) [reply package-lock.json:3085]: I guess is just because `typescript` is now in DEV section of package.json 

Looks ok for me.

**Review by mrseanryan — COMMENTED** [review]
@sverweij looks good - works for me on biggish project.

Can you add a line to the CHANGELOG ?

Also could be good to mention requires `typescript` as a *peer* dependency (can mention this at end of the Usage section).

**Comment by mrseanryan** [comment]
@sverweij looks good - works for me on biggish project.

Can you add a line to the CHANGELOG ?

Also could be good to mention requires `typescript` as a *peer* dependency (can mention this at end of the Usage section).

**Comment by sverweij** [comment]
Hi @mrseanryan thanks for the review - I've added some copy to both the changelog and the readme, trying to match the tone of voice of the surrounding material. That said - copy writing is highly subjective and on top of that I'm not a native speaker of English - so feel free to make amendments as you see fit.

**Comment by mrseanryan** [comment]
@sverweij  looks good!

Can you  rebase on master? (conflict in CHANGELOG)

Otherwise, I can rebuild the PR (we might need to do that anyway, due to github/travis issue...)

@pzavolinsky  I propose we first release `master` as `6.3.0` and then merge this PR, releasing as `7.0.0`

**Comment by mrseanryan** [comment]
hi @sverweij  again - actually no point in rebasing at this time, until we release from master

Maybe @pzavolinsky will take care of that anyway, when releasing 😁

**Comment by sverweij** [comment]
@mrseanryan rebased anyway before I noticed you second message - but I now see what you mean 😄. Should be easy to correct again, though.

**Comment by pzavolinsky** [comment]
:wave: @sverweij @mrseanryan rebased, fixed conflcits and meged this PR [here](https://github.com/pzavolinsky/ts-unused-exports/commit/9fee16e2531fb1aefeb18be3785d09ab8fdaf8b5).

Thanks a lot for this contrib!

**Pull request closed without merging**